src/main/java/ch/njol/skript/lang/parser/ParsingStack.java

```diff
@@ -0,0 +1,222 @@
+	 *
+	 * @throws IndexOutOfBoundsException if the given index does not point to
+	 * 										an element in the stack.
+	 */
+	public Element peek(int index) throws IndexOutOfBoundsException {
+		if (index < 0 || index >= size()) {
+			throw new IndexOutOfBoundsException("Index: " + index);
+		}
+
+		return stack.get(index);
+	}
+
+	/**
+	 * Returns the top element of the stack.
+	 * Equivalent to {@code peek(0)}.
+	 *
+	 * @throws IllegalStateException if the stack is empty.
+	 */
+	public Element peek() throws IllegalStateException {
+		if (stack.isEmpty()) {
+			throw new IllegalStateException("Stack is empty");
+		}
+
+		return stack.peek();
+	}
+
+	/**
+	 * Adds the given element to the top of the stack.
+	 */
+	public void push(Element e) {
+		stack.push(e);
+	}
+
+	/**
+	 * Check if this stack is empty.
+	 */
+	public boolean isEmpty() {
+		return stack.isEmpty();
+	}
+
+	/**
+	 * Gets the size of the stack.
+	 */
+	public int size() {
+		return stack.size();
+	}
+
+	/**
+	 * Prints this stack to the given {@link PrintStream}.
+	 */
+	public void print(PrintStream printStream) {
+		// Synchronized to assure it'll all be printed at once,
+		//  PrintStream uses synchronization on itself internally, justifying warning suppression
+
+		//noinspection SynchronizationOnLocalVariableOrMethodParameter
+		synchronized (printStream) {
+			printStream.println("Stack:");
+
+			if (stack.isEmpty()) {
+				printStream.println("<empty>");
+			} else {
+				for (Element element : stack) {
+					printStream.println("\t" + element.getSyntaxElementClass().getName()
+						+ " @ " + element.getPatternIndex());
+				}
+			}
+		}
+	}
+
+	/**
+	 * Iterate over the stack, starting at the top.
+	 */
+	@Override
+	public Iterator<Element> iterator() {
+		Iterator<Element> iterator = stack.iterator();
+
+		// Wrap iterator to disable element removal support
+		return new Iterator<Element>() {
+			@Override
+			public boolean hasNext() {
+				return iterator.hasNext();
+			}
+
+			@Override
+			public Element next() {
+				return iterator.next();
+			}
+		};
+	}
+
+	/**
+	 * A stack element, containing details about the syntax element it is about.
+	 */
+	public static class Element {
+
+		private final SyntaxElementInfo<?> syntaxElementInfo;
+		private final int patternIndex;
+
+		public Element(SyntaxElementInfo<?> syntaxElementInfo, int patternIndex) {
+			assert patternIndex >= 0 && patternIndex < syntaxElementInfo.getPatterns().length;
+
+			this.syntaxElementInfo = syntaxElementInfo;
+			this.patternIndex = patternIndex;
+		}
+
+		/**
+		 * Gets the raw {@link SyntaxElementInfo} of this stack element.
+		 * <p>
+		 * For ease of use, consider using other getters of this class.
+		 *
+		 * @see #getSyntaxElementClass()
+		 * @see #getPattern()
+		 */
+		public SyntaxElementInfo<?> getSyntaxElementInfo() {
+			return syntaxElementInfo;
+		}
+
+		/**
+		 * Gets the index to the registered patterns for the syntax element
```

Comment:
```suggestion
		 * Gets the index to the registered patterns for the syntax element info
```

Reply:
I disagree, the SyntaxElementInfo API is mostly unused by addon devs and should therefore not be mentioned in this doc. I added the util methods for directly getting the class, pattern or the like for this reason.

Reply:
It's coming from a SyntaxElementInfo instance

Reply:
Which addon devs don't use or know